### DISCUSSION TITLE
I would really appreciate an API for this package that could power a web UI or general programatic usage

### DISCUSSION THREAD

**Discussion by vesper8** [stefanzweifel/laravel-backup-restore#69] [Ideas]
I know we can always use the Artisan helper to call these commands from controllers but it would be really nice if you included a Facade or other helper and provided documentation on how to make use of all functionalities without ever depending on Artisan or running things from the console.

I already forked and hacked it up a bit so that I could sort of get it to power a web UI but it would be great if this was provided directly.

Many thanks for considering!

**Comment by stefanzweifel** [comment]
Cool idea!

A web UI is not something I have planned for this package, but if you've built something and would like others to use it as well, feel free to share your code as a package or as a PR to this repo. (I can't guarantee a merge)

What would you expect in a web UI? A list of all recent backups and a button to restore the backup? 
The package could support this, but that would be quite the time investment. A simple controller/view setup wouldn't work, as HTTP requests usually timeout after 30 seconds. (Backups I regularly restore take 5-15 minutes).
So we would have to add queued jobs and a way track progress on a restore.

Could you maybe explain in which context you would use such a UI? I hopefully assume I would run the UI on your local machine to restore a backup for the app your working on?

---

As you probably noticed, the core class in this package is [PendingRestore](https://github.com/stefanzweifel/laravel-backup-restore/blob/main/src/PendingRestore.php).
And the rest of the functionality has been moved into dedicated [Action classes](https://github.com/stefanzweifel/laravel-backup-restore/tree/main/src/Actions).

I ususally don't document the internal structure of a package, as that's subject to change and I don't want to create new major versions, just because I renamed an internal method.

---

I recently came across this Symfony package called [dbtoolsbundle](https://dbtoolsbundle.readthedocs.io/en/stable/), that has a similar feature set as this package. Maybe we can incorporate some of the features in a future release.

**Reply by vesper8** [reply]
Thanks for the quick response : )

Yes I also just found dbtoolsbundle. I even tried to use it in a Laravel project but couldn't figure out how to. My current project makes use of https://github.com/archtechx/tenancy and I'm building a web UI to allow users to do db backup/restore themselves as well as db duplicating/cloning which involves creating a backup of a tenant, creating a new tenant and restoring that backup onto the new tenant. In my use-case the DBs are small and they should always remain relatively small. I hadn't considered that in the far future some of them could potentially be taking more than a minute to restore.

As my use-case is a bit complex I opted to use Spatie\DbDumper directly instead of Spatie's backup package. I then struggled to find a solid existing solution for restoring a backup that allows me to specify both the dump filepath as well as the db name/credentials to restore to. That's when I found your package. I felt that your package does a lot of what i want but lacks the flexibility to let me do all of what I want.. so I started looking at using the underlying abstractions and that's when I decided to put that on hold, write this discussion and for now I'm just using this code to restore my backups:

```

    public function restoreBackupTo($product)
    {
        $command = sprintf(
            'mysql -u%s -p%s %s < %s',
            $product->tenant->tenancy_db_username,
            $product->tenant->tenancy_db_password,
            $product->tenant->tenancy_db_name,
            Storage::path($this->path)
        );

        exec($command);
    }
```

I also did create a super basic proof-of-concept web UI by forking your package and adding a new --format option that can return the list of backups as json:

https://github.com/vesper8/laravel-backup-restore/blob/a810982f51ae8b83b270e5c59a3b0b1e68ced1ff/src/Commands/RestoreCommand.php#L68

I then make use of this controller

```

<?php

namespace App\Http\Controllers;

use Artisan;
use Illuminate\Http\Request;
use App\Http\Controllers\Controller;

class BackupController extends Controller
{
    public function index()
    {
        // Run the Artisan command to get backups
        $output = Artisan::call('backup:restore', [
            '--format' => 'json',
        ]);

        // Get the output and decode it to get backups as an array
        $backups = json_decode(Artisan::output(), true);

        // Return the view with backups
        return view('backups.index', compact('backups'));
    }


    public function restore(Request $request)
    {
        // Validate the request
        $request->validate([
            'backup-name' => 'required|string',
        ]);

        // Run the Artisan command to restore a backup
        $backupName = $request->input('backup-name');

        $options = [
            '--backup' => $backupName,
            '--quiet' => true,
        ];

        Artisan::call('backup:restore', $options);

        // Redirect back with a success message
        return back()->with('success', 'Backup restored successfully.');
    }
}
```

And this blade template

```

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Backups</title>
    <link href="{{ asset('css/app.css') }}" rel="stylesheet">
</head>
<body class="bg-gray-100 p-8">
    <div class="container mx-auto">
        <h1 class="text-xl mb-4">Available Backups</h1>
        @if(session('success'))
            <div class="bg-green-100 border border-green-400 text-green-700 px-4 py-3 rounded relative" role="alert">
                {{ session('success') }}
            </div>
        @endif
        <div class="bg-white shadow-md rounded px-8 pt-6 pb-8 mb-4">
            <ul class="list-disc">
                @foreach($backups as $backup)
                    <li>
                        <form action="{{ route('backups.restore') }}" method="POST">
                            @csrf
                            <input type="hidden" name="backup-name" value="{{ $backup }}">
                            <button type="submit" class="text-blue-500 hover:text-blue-700">{{ $backup }}</button>
                        </form>
                    </li>
                @endforeach
            </ul>
        </div>
    </div>
</body>
</html>
```

Very basic stuff and does not at all handle restorations that take a long time.

**Reply by stefanzweifel** [reply]
Sorry for my late reply. 🫠

Great that you could re-use the underlying Action classes.

Someone else mentioned in #52, that they use _a_ multi-tenancy package and would like to use this package in their project.
I honestly currently can't think of a good solution for this problem.

How a database (or project file #51) is being restored can differ from project to project. Especially in projects that use multi-tenancy with separate databases.

I currently don't have the capacity to build this.
In the last 10 years I never had to support multi-tenancy through separate databases, so I don't really have any experience on how I would use the package.

However, I'm currently also not really happy on how database dumps are being imported in this project. There's definitely room for improvement.

When writing the package, I was also looking for other packages that solved this problem … and couldn't find one. So I built it myself.
But `dbtoolsbundle` looks nice. I made a note to check that package out in more detail and maybe repurpose some of their internal code/packages in this package as well.

**Comment by stefanzweifel** [comment]
Closing this discussion as it’s out of scope for this package.

I don't use this package through a web UI and don’t see a need for that functionality in the future.

Going into 2026 and beyond, I’m intentionally limiting new development to features I actively need or that clearly benefit the package as a whole.

If this functionality is important to you, feel free to fork the package and implement it yourself. A well-thought-out pull request is welcome, and I’m happy to review it when time allows.
